### ISSUE TITLE
DynamicNode: implement Rust `get_` methods to get dynamic node properties.

### ISSUE THREAD

**Issue by inickles-grapl** [grapl-security/grapl#233]
### What did you want to do?

I want to implement something like the following: 

```
fn try_from(mut my_node: MyNode) -> Result<IpAddress, failure::Error> {
    ....
}
```

But it's not immediately obvious how to get properties previously set from a dynamic node. The Rust definition for DynamicNode is generated dynamically from a mix of Rust and protobuf.

### What did you actually end up doing instead?

I asked Colin for help :)

```
    fn try_from(mut my_node: MyNode) -> Result<IpAddress, failure::Error> {
        let dynamic_node: &mut DynamicNode = my_node.get_mut_dynamic_node();

        let property_name = "source_ip_address";
        let source_ip_address = dynamic_node
            .properties
            .get(property_name)
            .ok_or_else(|| format_err!("Node property not found: {}", property_name))?
            .as_str_prop()
            .ok_or_else(|| format_err!("Node property not str: {}", property_name))?;

        let property_name = "timestamp";
        let timestamp = dynamic_node
            .properties
            .get(property_name)
            .ok_or_else(|| format_err!("Node propery not found: {}", property_name))?
            .as_int_prop()
            .ok_or_else(|| format_err!("Node property not str: {}", property_name))?;

        IpAddressBuilder::default()
            .ip_address(source_ip_address)
            .last_seen_timestamp(timestamp as u64)
            .build()
            .map_err(|s| format_err!("{}", s))
    }
```

### What can Grapl do differently?

Similar to how dynamic nodes have `with_` methods to set node properties, proved `get_` methods that returns property values. Something like:

```
    fn try_from(mut my_node: MyNode) -> Result<IpAddress, failure::Error> {
        let dynamic_node: &mut DynamicNode = my_node.get_mut_dynamic_node();

        let source_ip_address = dynamic_node
            .get_source_ip_address()
            .ok_or_else(|| format_err!("Node property not found: source_ip_address"))?

        let property_name = "timestamp";
        let timestamp = dynamic_node
            .get_timestamp()
            .ok_or_else(|| format_err!("Node property not found: timestemp"))?

        IpAddressBuilder::default()
            .ip_address(source_ip_address)
            .last_seen_timestamp(timestamp as u64)
            .build()
            .map_err(|s| format_err!("{}", s))
    }
```

**Comment by alexjmacdonald** [comment]
I'm working on an implementation for this, and about 90% of the way there.  I hit one issue that I'd like to raise here:

```
fn #get_method_name(&self) -> Option<&#property_type> {
    // properties hashmap returns an Option<&NodeProperty>, bc we can't guarantee key is present
    let property_result: Option<&NodeProperty> = self.get_dynamic_node().get_property(#property_name_str);
    // T::from(NodeProperty) turns an Option<T> bc we can't guarantee NodeProperty holds a T
    match property_result {
        Some(property) => Option::< #property_type >::from(property),
        None => None
    }
}
```

Without any finessing, the return type would be `Option<Option<#property_type>>`, because:
1) `properties` isn't guaranteed to have the key
2) the value isn't guaranteed to convert cleanly into type T.

This makes the method really inconvenient to use.

So, I'm thinking of a few API options here for `get_#property_name`:
1)  `get_#property_name` returns `Option<#property_type>`, where None indicates _any_ issue in retrieving the property.
2)  We assume dynamic properties are always semantically valid, and `panic!` if a property is missing or badly typed.

I'm leaning toward #2, given the example struct:

```
#[derive(DynamicNode, GraplStaticId)]
pub struct AwsEc2Instance {
    #[grapl(static_id)]
    arn: String,
    #[grapl(static_id)]
    launch_time: u64,
}
```

This has `String` and `u64` as fields, not Options.  I would assume we would carry the same invariants over to the dynamic properties?

**Comment by colin-grapl** [comment]
Interesting. I think we would probably want to still return Option<#property_type>, but unwrap the "well formed" type. I think it's an OK invariant to say "well formed", especially in the graph-generators, but it's unexpected for 'get' to panic on a missing propery.

Or, alternatively, we could have another enum, like:
```
pub enum DynamicProperty<T> {
    Some(T),
    Missing,
    Malformed,
}
```
But that's a bit weird/ throws away tons of helper methods that Option provides.

It's a bit of a wart of the Derive macro that fields aren't optional, but *they are* actually optional. The only ones that aren't are the ones marked for static_id.

**Comment by wimax-grapl** [comment]
Closes #378
